### PR TITLE
keep an eventlet-friendly reference to time.sleep

### PR DESCRIPTION
**Pull request by davidszotten**
otherwise, when eventlet.monkey_patch replaces time.sleep,
this raises a TypeError when called, since while adding the builtin
sleep to the ReloaderLoop we stil end up with a function, but when
replaced, we end up with a method, which now expects a self argument.

### CODE REVIEW

**Comment by untitaker** [comment]
That's reasonable. However, I am not sure how eventlet would e.g. deal with

    from time import sleep

if it can't even deal with this case.

Anyway, this patch is fine. Could you add an additional comment explaining the situation, basically the one that you wrote in the PR?

**Comment by davidszotten** [comment]
added comment, though rather lengthy. will happily take suggestions for a better, briefer way to explain this

File: tests/test_serving.py [hunk 1]
```diff
@@ -171,3 +172,21 @@ def real_app(environ, start_response):
...
+        pass
+
+    # simulate eventlet.monkey_patch by replacing the builtin sleep
+    # with a regular function
+    time.sleep = sleep
+
+    from werkzeug._reloader import ReloaderLoop
+    ReloaderLoop()._sleep(0)
+    '''))
+    subprocess.check_call(['python', str(script)])
```
Comment (untitaker) [inline tests/test_serving.py]: I don't think this test is useful, as all it does is sleep when the method didn't get patched properly.

Also, you can safely do monkeypatching with [pytest](http://pytest.org/latest/monkeypatch.html), no need to spawn a new process.

File: werkzeug/_reloader.py [hunk 2]
```diff
@@ -107,9 +107,19 @@ def _walk(node, path):
...
-    _sleep = time.sleep  # monkeypatched by testsuite
+
     name = None
 
+    # monkeypatched by testsuite. this used to just say `_sleep = time.sleep`,
+    # but because c functions behave differently to python functions when
+    # attached to classes, this breaks if sleep has been already replaced with
+    # a python function (e.g. by eventlet.monkey_patch). python functions are
+    # descriptors and become methods when attahed to classes, whereas c
+    # functions remain as functions (thus behaving like staticmethods)
```
Comment (untitaker) [inline werkzeug/_reloader.py]: Just:

    Use a wrapper method instead of a simple assignment to allow monkeypatching by e.g. eventlet

File: tests/test_serving.py (hunk 1 above)
Comment (untitaker) [inline tests/test_serving.py]: On further thought, I don't think the patch is complex enough to warrant a testcase.

Comment (davidszotten) [inline tests/test_serving.py]: this isn't testing that the patching works, but rather that the copy of `sleep` is attached "safely". if you revert the change in `_reloader.py` and run this test you see the issue i'm trying to fix

File: werkzeug/_reloader.py (hunk 2 above)
Comment (davidszotten) [inline werkzeug/_reloader.py]: it's not monkeypatching that fails, it's _using_ `ReloaderLoop._sleep` if `sleep` was monkeypatched already _before_ it was attached to the `ReloaderLoop`

File: tests/test_serving.py (hunk 1 above)
Comment (untitaker) [inline tests/test_serving.py]: Ok, that was a fundamental misunderstanding on my side on what this patch is trying to fix.

In that case, might

    _sleep = staticmethod(time.sleep)

not work as well?

File: werkzeug/_reloader.py (hunk 2 above)
Comment (untitaker) [inline werkzeug/_reloader.py]: Ok, got it.

**Comment by untitaker** [comment]
This looks good, sorry for the confusion.

**Comment by untitaker** [comment]
Thanks!

File: werkzeug/_reloader.py (hunk 2 above)
Comment (davidszotten) [inline werkzeug/_reloader.py]: clearly my comment doesn't do a great job of explaining this though :(  could have a briefer comment and reference this thread perhaps?

Comment (untitaker) [inline werkzeug/_reloader.py]: Probably inattention on my part, no worries.

File: tests/test_serving.py (hunk 1 above)
Comment (davidszotten) [inline tests/test_serving.py]: ah yes. that's nicer

Comment (untitaker) [inline tests/test_serving.py]: Nevermind, I find yours easier to read.
